dali/operators/decoder/audio/utils.h

```diff
@@ -0,0 +1,168 @@
+ * Downmixing data to mono-channel. It's safe to do it in-place.
+ *
+ * Assumes, that `out` buffer is properly allocated
+ * (i.e. has `n_in`/`weights.size()` elements at least)
+ *
+ * @param weights Specifies weight for every channel.
+ */
+template<typename T>
+void Downmixing(T *out, const T *in, size_t n_in, const std::vector<int> &weights) {
+  DALI_ENFORCE(out && in, "Incorrect input or output buffer. Or both.");
+  DALI_ENFORCE(weights.size() > 1, "You can't downmix mono-channel data");
+  DALI_ENFORCE(n_in % weights.size() == 0, "Data layout incorrect");
+  for (size_t i = 0; i < n_in; i++) {
+    *out = (*in++) * weights[0];
+    for (size_t j = 1; j < weights.size(); j++) {
+      *out += (*in++) * weights[j];
+    }
+    *out++ /= std::accumulate(weights.begin(), weights.end(), 0);
+  }
+}
+
+
+template<typename T>
+void Downmixing(T *out, const T *in, size_t n_in, int n_channels_in) {
+  Downmixing(out, in, n_in, {n_channels_in, 1});
+}
+
+
+template<typename T>
+void Downmixing(span<T> out, span<const T> in, const std::vector<int> &weights) {
+  Downmixing(out.data(), in.data(), in.size(), weights);
+}
+
+
+template<typename T>
+void Downmixing(span <T> out, span<const T> in, size_t n_channels_in) {
+  Downmixing(out.data(), in.data(), in.size(), n_channels_in);
+}
+
+namespace resampling {
+double Hann(double x) {
+  return 0.5 * (1 + std::cos(x * M_PI_2));
+}
+
+
+double sinc(double x) {
+  x *= M_PI;
+  if (std::abs(x) < 1e-10)
+    return 1 - x * x * 0.25;  // approximate by a parabola near the pole
+  return std::sin(x) / x;
+}
+
+
+struct sinc_coeffs {
+  void init(int coeffs, int lobes = 3, std::function<double(double)> envelope = Hann) {
+    float scale = 2.0f * lobes / (coeffs - 1);
+    float scale_envelope = 2.0f / coeffs;
+    this->coeffs = coeffs;
+    this->lobes = lobes;
+    lookup.resize(coeffs + 2);  // add zeros
+    center = (coeffs - 1) * 0.5f;
+    for (int i = 0; i < coeffs; i++) {
+      float x = (i - center) * scale;
+      float y = (i - center) * scale_envelope;
+      float w = sinc(x) * envelope(y);
+      lookup[i + 1] = w;
+      std::cerr << i << ": " << w << "\n";
+    }
+    center++;  // allow for leading zero
+    this->scale = 1 / scale;
+  }
+
+
+  std::pair<int, int> input_range(float x) const {
+    int i0 = std::ceil(x) - lobes;
+    int i1 = std::floor(x) + lobes;
+    return {i0, i1};
+  }
+
+
+  float operator()(float x) const {
+    float fi = x * scale + center;
+    int i = std::floor(fi);
+    float di = fi - i;
+    assert(i >= 0 && i < (int) lookup.size());  // NOLINT
+    return lookup[i] + di * (lookup[i + 1] - lookup[i]);
+  }
+
+
+  float scale = 1, center = 1;
+  int lobes = 0, coeffs = 0;
+  std::vector<float> lookup;
+};
+
+
+/**
+ * Resampling for audio buffer
+ */
+void resample_sinc(
+        float *out, int64_t n_out, double out_rate,
+        const float *in, int64_t n_in, double in_rate,
+        int lobes = 3) {
+  sinc_coeffs coeffs;
+  coeffs.init(1024, lobes);
```

Comment:
One node on this function: the code was a quick-and-dirty prototype that would just do the job and keep reasonable precision. There are shortcomings, not least of which is the fact that the sinc lookup is recalculated on each call. It'd be far better to rework it into a SincResampler class, which would maintain the lookup table across calls.